kalamine/data/key_codes.yaml

```diff
@@ -1,65 +1,87 @@
-klc:
-  spce: '39	SPACE'
-
-  # digits
-  ae01: '02	1'
-  ae02: '03	2'
-  ae03: '04	3'
-  ae04: '05	4'
-  ae05: '06	5'
-  ae06: '07	6'
-  ae07: '08	7'
-  ae08: '09	8'
-  ae09: '0a	9'
-  ae10: '0b	0'
-
-  # letters, first row
-  ad01: '10	Q'
-  ad02: '11	W'
-  ad03: '12	E'
-  ad04: '13	R'
-  ad05: '14	T'
-  ad06: '15	Y'
-  ad07: '16	U'
-  ad08: '17	I'
-  ad09: '18	O'
-  ad10: '19	P'
-
-  # letters, second row
-  ac01: '1e	A'
-  ac02: '1f	S'
-  ac03: '20	D'
-  ac04: '21	F'
-  ac05: '22	G'
-  ac06: '23	H'
-  ac07: '24	J'
-  ac08: '25	K'
-  ac09: '26	L'
-  ac10: '27	OEM_1'
-
-  # letters, third row
-  ab01: '2c	Z'
-  ab02: '2d	X'
-  ab03: '2e	C'
-  ab04: '2f	V'
-  ab05: '30	B'
-  ab06: '31	N'
-  ab07: '32	M'
-  ab08: '33	OEM_COMMA'
-  ab09: '34	OEM_PERIOD'
-  ab10: '35	OEM_2'
-
-  # pinky keys
-  tlde: '29	OEM_3'
-  ae11: '0c	OEM_MINUS'
-  ae12: '0d	OEM_PLUS'
-  ae13: '0d	OEM_102' # XXX FIXME
-  ad11: '1a	OEM_4'
-  ad12: '1b	OEM_6'
-  ac11: '28	OEM_7'
-  ab11: '28	OEM_7'  # XXX FIXME
-  bksl: '2b	OEM_5'
-  lsgt: '56	OEM_102'
+klc: # more info on SC and VK: https://learn.microsoft.com/en-us/windows/win32/inputdev/about-keyboard-input
+  sc: # Scan codes. They are the code sent by the hardware
+    spce: '39'
+
+    # digits
+    ae01: '02'
+    ae02: '03'
+    ae03: '04'
+    ae04: '05'
+    ae05: '06'
+    ae06: '07'
+    ae07: '08'
+    ae08: '09'
+    ae09: '0a'
+    ae10: '0b'
+
+    # letters, first row
+    ad01: '10'
+    ad02: '11'
+    ad03: '12'
+    ad04: '13'
+    ad05: '14'
+    ad06: '15'
+    ad07: '16'
+    ad08: '17'
+    ad09: '18'
+    ad10: '19'
+
+    # letters, second row
+    ac01: '1e'
+    ac02: '1f'
+    ac03: '20'
+    ac04: '21'
+    ac05: '22'
+    ac06: '23'
+    ac07: '24'
+    ac08: '25'
+    ac09: '26'
+    ac10: '27'
+
+    # letters, third row
+    ab01: '2c'
+    ab02: '2d'
+    ab03: '2e'
+    ab04: '2f'
+    ab05: '30'
+    ab06: '31'
+    ab07: '32'
+    ab08: '33'
+    ab09: '34'
+    ab10: '35'
+
+    # pinky keys
+    tlde: '29'
+    ae11: '0c'
+    ae12: '0d'
+    ae13: '0d' # XXX FIXME
+    ad11: '1a'
+    ad12: '1b'
+    ac11: '28'
+    ab11: '28'  # XXX FIXME
+    bksl: '2b'
+    lsgt: '56'
+  vk: # Virtual keys
+    # the driver maps scan codes to VK which are in turn 
+    # translated by the layout to output a char.
+    # We don't put digits nor alpha chars as they are their own VK
+    '2019': 'OEM_1' # 1dk
```

Comment:
I used the values here https://learn.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes
Then added special case `U+2019` for Ergo-L

I used `OEM_1` for `U+2019` because it was what I’ve seen in https://github.com/Nuclear-Squid/ergol/blob/b03d43e33cfd5d03a1adb600aa95ba4390841b61/0_7_0/ergol.klc#L79

If it’s not correct, I’m don’t know how to assign correct VK then :-/  

You talked about a complex algorithm to manage it, maybe that’s the way we need?

Reply:
Ok I think I get the idea of where we need to go from there and it’s what you said in comment below, so let’s close this one